google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/wrappers/Range.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.wrappers;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+
+/**
+ * Range API.
+ *
+ * <p>This base class represents the API for all ranges in the Cloud Bigtable client.
+ */
+abstract class Range<T, R extends Range<T, R>> {
+  public enum BoundType {
+    OPEN,
+    CLOSED,
+    UNBOUNDED
+  }
+
+  final BoundType startBound;
+  final T start;
+  final BoundType endBound;
+  final T end;
+
+  Range() {
+    this(BoundType.UNBOUNDED, null, BoundType.UNBOUNDED, null);
+  }
+
+  Range(BoundType startBound, T start, BoundType endBound, T end) {
+    this.start = start;
+    this.startBound = startBound;
+    this.end = end;
+    this.endBound = endBound;
+  }
+
+  /**
+   * Creates a new {@link Range} with the specified inclusive start and the specified exclusive end.
+   */
+  public R of(T startClosed, T endOpen) {
```

Comment:
public static?

Reply:
Yea, it has to be an instance method to allow chainability:

FILTERS.timestamp().range().of(...)